Сделана вторая часть с имплементацией сервисов в main (week1)

crafty-ezhik opened the pull request.

crafty-ezhik committed:
Change Taskfile

crafty-ezhik committed:
Сделано описание для Order API

Описаны необходимые ручки, ошибки, enums, ответы, payloads в формате OpenAPI

crafty-ezhik committed:
Сгенерирован код API для Order Service

crafty-ezhik committed:
feat: Добавлена обработка неизвестных ошибок

В paths добавлена секция default для не типичных ошибок

crafty-ezhik committed:
feat: Написан proto файл для InventoryService

crafty-ezhik committed:
fix: Удалены неиспользуемые импорты

crafty-ezhik committed:
feat: Написан proto файл для PaymentService

crafty-ezhik committed:
fix: Изменены названия в enum proto файлах

crafty-ezhik committed:
feat: Сгенерирован код из proto файлов

Сгенерирован код для InventoryService и PaymentService. Также в настройки линтера proto файла добавлено исключение ENUM_VALUE_PREFIX, иначе в go получается дублирование, к примеру "PaymentMethod_PAYMENT_METHOD_CARD", что выглядит нечитаемо.

crafty-ezhik committed:
fix: поменян префис для ипортов в taskfile

crafty-ezhik committed:
feat: Добавлено описание в proto для gRPC gateway

crafty-ezhik committed:
feat: Сгенерирован код для gRPC gateway

Сгенерирован код для InventoryService и PaymentService

crafty-ezhik committed:
fix: Изменен taskfile

crafty-ezhik committed:
fix: Поменяно расположение сгенерированного кода

Сгенерированный код перемещен в shared/pkg. До этого было просто в pkg в корне проекта

crafty-ezhik committed:
feat: Добавлена валидация в proto файлах

crafty-ezhik committed:
feat: Сгенерирован код для валидации

crafty-ezhik committed:
feat: Добавлена генерация swagger из proto

crafty-ezhik committed:
feat: Добавлена генерация Swagger для сервисов

Добавлена генерация swagger для InventoryService и PaymentService

crafty-ezhik committed:
feat: заготовки под реализацию логики

crafty-ezhik committed:
Merge pull request #2 from crafty-ezhik/hw-1_part-2

Добавлены файлы по кодогенерации

crafty-ezhik committed:
feat: базово имплементирован PaymentService

Также исправлены мелкие косяки, чтобы линтер не ругался. Создана задача для упрощения форматирования и линтинга кода.

crafty-ezhik committed:
feat: Добавлен gateway и swagger для PaymentService

crafty-ezhik committed:
feat: добавлены интерцепторы в PaymentService

Добавлено логирование и валидация входных параметров. Вынесено в интерцепторы

crafty-ezhik committed:
style: Форматирование кода

crafty-ezhik committed:
feat: базово имплементирован InventoryService

crafty-ezhik committed:
feat: Реализован метод GetPart

Также отформатирован код + отсортированы импорты

crafty-ezhik committed:
feat: Реализован метод ListParts

Фильтрация реализована через generic для string и Category. []string фильтруется отдельно, чтобы не усложнять generic функцию

crafty-ezhik committed:
fix: Исправлена openAPI схема

Исправлено представление part_uuids c []jx.Raw на []uuid.UUID, total_price с float(float32) на double(float64)

crafty-ezhik committed:
fix: Исправление openAPI файла

Поменян тип с float на double

crafty-ezhik committed:
fix: Исправлена ошибка out of range

crafty-ezhik committed:
refactor: Поправлена генерация данных

crafty-ezhik committed:
fix: Исправлен формат ответа на GetOrder

crafty-ezhik committed:
feat: Реализована логика для создания и получения заказа

crafty-ezhik committed:
fix: Исправлен ответ при отмене заказа в 204 коде

crafty-ezhik committed:
feat: Реализована логика отмены и оплаты заказа

crafty-ezhik committed:
fix: Правки для того, чтобы тесты проходили корректно

crafty-ezhik committed:
chore: добавлены комментарии

crafty-ezhik committed:
fix: Исправления после ревью

Исправлена опечатка, связанная с UNKNOWN. Также поля transaction_uuid/payment_method сделаны необязательными. Исправлена логика в OrderService в связи с изменением обязательности полей transaction_uuid/payment_method.

crafty-ezhik committed:
fix: Исправлен путь

mistandok commented on inventory/cmd/main.go:
```diff
@@ -0,0 +1,314 @@
+)
+
+// mapParts -псевдоним для map[string]*inventoryV1.Part
+type mapParts map[string]*inventoryV1.Part
```
issue: в гошке лучше не создавать свои типы поверх map, slice и т.п., можно неожиданное поведение словить. Например создашь тип на основе мапы, кто-то не посмотрит и будет в конкурентном коде пытаться использовать, словит панику. Просто инкапсулируй внутри сервиса, разницы никакой не будет.

mistandok commented on inventory/cmd/main.go:
```diff
@@ -0,0 +1,314 @@
+
+// applyFilter - применяет фильтр к мапе. Передается список значений фильтра по которому фильтровать
+// и геттер для получения значения у n-го элемента
+func applyFilter[T FilterType](mp mapParts, filter []T, getter func(p *inventoryV1.Part) T) {
```
issue: у тебя тут каждый фильтр будет проходить по всему списку запчастей. а это m * n сложность, где m - количество запчастей, а n - количество фильтров. 

Сделай проще, через замыкание и без дженериков, читаться будет легче

```go
	type filter func(part *schema.Part) bool

	filters := []filter{
		func(part *schema.Part) bool {
			if len(uuidSet) > 0 {
				if _, ok := uuidSet[part.Uuid]; !ok {
					return false
				}
			}
			return true
		},
		func(part *schema.Part) bool {
			if len(nameSet) > 0 {
				if _, ok := nameSet[strings.ToLower(part.Name)]; !ok {
					return false
				}
			}
			return true
		},
	}

	var filtered []*schema.Part
	for _, part := range parts {
		for _, filter := range filters {
			if filter(part) {
				filtered = append(filtered, part)
				break
			}
		}
	}
```

mistandok commented on order/cmd/main.go:
```diff
@@ -0,0 +1,435 @@
+}
+
+// PayOrder - меняет статут оплаты, метод оплаты и уникальный идентификатор транзакции
+func (s *OrderStorage) PayOrder(orderUUID, transactionUUID uuid.UUID, paymentMethod orderV1.PaymentMethod) error {
```
suggestion: если это storage, то на storage лучше сделать метод updateOrder. PayOrder - это бизнес логика, а не работа с хранилищем.

В целом, давай пока не будем сюда лезть, оставь как есть, посмотришь вторую неделю, станет понятнее

mistandok commented on order/cmd/main.go:
```diff
@@ -0,0 +1,435 @@
+// OrderCreate - Создает заказ и возвращает его уникальный идентификатор
+func (h *OrderHandler) OrderCreate(ctx context.Context, req *orderV1.CreateOrderRequest) (orderV1.OrderCreateRes, error) {
+	// Валидируем запрос
+	if err := req.Validate(); err != nil {
```
suggestion: вынеси валидации входящих данных в отдельную функцию, проще код будет читаться. с 166 по 179 срочку

mistandok commented on order/cmd/main.go:
```diff
@@ -0,0 +1,435 @@
+	}
+
+	// Преобразуем входящий []uuid.UUID к []string
+	partsUUID := make([]string, 0, len(req.GetPartUuids()))
```
suggestion: ты пишешь комментарии по той причине, что хочешь рассказать, что делает код. Зачем писать комментарий, если можно код выделить в функцию с говорящим названием? convertUUIDSToString() и все

mistandok commented on order/cmd/main.go:
```diff
@@ -0,0 +1,435 @@
+	}
+
+	// Идем в InventoryService для получения списка запрашиваемых деталей
+	parts, err := h.inventoryClient.ListParts(ctx, &inventoryV1.ListPartsRequest{Filter: &inventoryV1.PartsFilter{
```
suggestion(blocking): всегда лучше для общения с клиентами задавать какой-то таймаут, иначе можно нарваться на долгое соединение и не понимать, почему же ручка висит. Это можно сделать либо на этапе инициализации клиента, если мы пишем свою обертку над ним или используем внешние библиотеки, которе позволяеют это делать, либо непосредственно в месте вызова через
```go
ctx := context.WithTimeout(ctx, 1 * time.Second)
```

Таймауты выставляем относительно NFR нашей ручки. Если мы гарантируем отдать ответ за 3 секунды, знаем, что наша БЛ работает 2 секунды, то на все сторонние вызовы у нас не более 1 секунды.

mistandok commented on order/cmd/main.go:
```diff
@@ -0,0 +1,435 @@
+		}, nil
+	}
+
+	// Проверяем, что все запрашиваемые детали есть в наличии и считаем total_price
```
suggestion: также можно в отдельную функцию

mistandok commented on order/cmd/main.go:
```diff
@@ -0,0 +1,435 @@
+	}
+
+	// Отправляем данные для оплаты в PaymentService
+	transactionUUIDstr, err := h.paymentClient.PayOrder(ctx, &paymentV1.PayOrderRequest{
```
issue(critical): можешь оплатить заказ в любом статусе, это неверно

mistandok reviewed (changes requested):
Привет! У тебя есть явное непонимание, что должно быть в storage, а что должно быть на уровне бизнес логики, но я эти моменты тут не отражал, потому что проходим это только на второй неделе. Там посмотрим, появилось ли это самое понимание.

В остальном пометил ошибки, которые на мой взгляд нужно исправить.

crafty-ezhik committed:
fix: Исправления

1. Убран псевдоним для map[string]*inventoryV1.Part
2. Изменено название функции для обновления заказов
3. Валидация вынесена в отдельную функцию
4. Преобразование []UUID в string вынесено в отдельную функцию
5. Проверка наличия деталей и подсчет итоговой стоимости вынесено в отдельную функцию

crafty-ezhik committed:
fix: Исправления

1. Добавлены таймауту при вызове gRPC сервисов
2. Добавлена проверка статуса заказа при оплате
3. Добавлен 409 код в качестве возможного ответа на запрос оплаты

crafty-ezhik committed:
perf: Оптимизирована фильтрация деталей

crafty-ezhik committed:
fix: Изменен возврат ошибок из функции

mistandok reviewed (approved):
пойдем ко второй неделе
